src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition2Test.java

```diff
@@ -71,6 +75,78 @@
         assertTrue(exec.isComplete());
     }
 
+    /**
+     * Verify that we kill endlessly recursive CPS code cleanly.
+     */
+    @Test
+    public void failureInNode() throws Exception {
+        String script = "node { assert 1 == 2; } ";
+        WorkflowJob job = jenkins.jenkins.createProject(WorkflowJob.class, "failInNode");
+        job.setDefinition(new CpsFlowDefinition(script, true));
+
+        // Should have failed with error about excessive recursion depth
+        WorkflowRun r = jenkins.assertBuildStatus(Result.FAILURE, job.scheduleBuild2(0).get());
+
+        Assert.assertTrue("No queued FlyWeightTask for job should remain after failure", jenkins.jenkins.getQueue().isEmpty());
+
+        for (Computer c : jenkins.jenkins.getComputers()) {
+            for (Executor ex : c.getExecutors()) {
+                if (ex.isBusy()) {
+                    fail(ex.getCurrentExecutable().toString());
+                }
+            }
+        }
+    }
+
+    /**
+     * Verify that we kill endlessly recursive CPS code cleanly.
+     */
+    @Test
+    public void endlessRecursion() throws Exception {
+        String script = "def getThing(){return thing == null}; \n" +
+                "node { echo getThing(); } ";
+        WorkflowJob job = jenkins.jenkins.createProject(WorkflowJob.class, "recursion");
+        job.setDefinition(new CpsFlowDefinition(script, true));
+
+        // Should have failed with error about excessive recursion depth
+        WorkflowRun r = jenkins.assertBuildStatus(Result.FAILURE, job.scheduleBuild2(0).get());
+        jenkins.assertLogContains("look for unbounded recursion", r);
+
+        Assert.assertTrue("No queued FlyWeightTask for job should remain after failure", jenkins.jenkins.getQueue().isEmpty());
+
+        for (Computer c : jenkins.jenkins.getComputers()) {
+            for (Executor ex : c.getExecutors()) {
+                if (ex.isBusy()) {
+                    fail(ex.getCurrentExecutable().toString());
+                }
+            }
+        }
+    }
+
+    /**
+     * Verify that we kill endlessly recursive NonCPS code cleanly and don't leave remnants.
```

Comment:
Is this just a native `StackOverflowError`?

Reply:
Yes, I wanted a little extra security.  Updated the comments.